File: BE-MyCarMaster/src/main/java/softeer/be_my_car_master/api/option/dto/response/OptionDto.java
package softeer.be_my_car_master.api.option.dto.response;

import java.util.List;
import java.util.stream.Collectors;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import softeer.be_my_car_master.domain.option.Option;

@Getter
@Setter
@Builder
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class OptionDto {

	@Schema(description = "옵션 식별자", example = "1")
	private Long id;

	@Schema(description = "카테고리", example = "SAFE")
	private String category;

	@Schema(description = "옵션명", example = "주차보조 시스템||")
	private String name;

	@Schema(description = "옵션 기본 비용", example = "1280000")
	private Integer price;

	@Schema(description = "판매율", example = "22")
	private Integer ratio;

	@Schema(description = "옵션 이미지", example = "null")
	private String imgUrl;

	@Schema(description = "옵션 요약", example = "null")
	private String summary;

	@Schema(description = "옵션 설명", example = "null")
	private String description;

	@Schema(description = "태그", example = "null")
	private String tag;

	private List<SubOptionDto> subOptions;

	public static OptionDto from(Option option) {
		List<SubOptionDto> subOptionDtos = null;
		if (option.hasSubOption()) {
			subOptionDtos = option.getSubOptions().stream()
				.map(SubOptionDto::from)
				.collect(Collectors.toList());
		}

		return OptionDto.builder()
			.id(option.getId())
			.category(option.getCategory())
			.name(option.getName())
			.price(option.getPrice())
			.ratio(option.getRatio())
			.imgUrl(option.getImgUrl())
			.summary(option.getSummary())
			.description(option.getDescription())
			.tag(option.getTagName())
			.subOptions(subOptionDtos)
			.build();
	}
}

File: iOS-MyCarMaster/MyCarMaster/MyCarMaster/Model/DataTransferObject/OptionDTO.swift
//
//  OptionDTO.swift
//  MyCarMaster
//
//  Created by SEUNGMIN OH on 2023/08/16.
//

import Foundation

struct OptionDTO: Codable {
    let id: Int
    let category: String
    let name: String
    let price: Int
    let ratio: Int
    let imgURL: String
    let summary: String
    let description: String
    let tag: String
    let subOptions: [SubOptionDTO]

    enum CodingKeys: String, CodingKey {
        case id = "id"
        case category = "category"
        case name = "name"
        case price = "price"
        case ratio = "ratio"
        case imgURL = "imgUrl"
        case summary = "summary"
        case description = "description"
        case tag = "tag"
        case subOptions = "subOptions"
    }
}

struct SubOptionDTO: Codable {
    let name: String
    let imgURL: String
    let description: String

    enum CodingKeys: String, CodingKey {
        case name = "name"
        case imgURL = "imgUrl"
        case description = "description"
    }
}

// MARK: Convert
extension Option {
    init(_ optionDTO: OptionDTO) {
        self.model = "펠리세이드"
        self.category = optionDTO.category
        self.name = optionDTO.name
        self.price = optionDTO.price
        self.ratio = optionDTO.ratio
        self.imgURL = URL(string: optionDTO.imgURL)
        self.summary = optionDTO.summary
        self.description = optionDTO.description
        self.tag = optionDTO.tag
        self.subOptions = optionDTO.subOptions.map { SubOption($0) }
    }
}

extension SubOption {
    init(_ subOptionDTO: SubOptionDTO) {
        self.name = subOptionDTO.name
        self.imgURL = URL(string: subOptionDTO.imgURL)
        self.description = subOptionDTO.description
    }
}
